DirectProgramming/DPC++FPGA/include/memory_tools.hpp

```diff
@@ -0,0 +1,69 @@
+#ifndef __MEMORY_TOOLS_HPP__
+#define __MEMORY_TOOLS_HPP__
+
+#include <type_traits>
+
+#include "metaprogramming_tools.hpp"
+
+namespace fpga_tools {
+
+template<typename Pipe, typename PtrT>
+void MemoryToPipe(PtrT in_ptr, size_t count_div_elements_per_cycle) {
+  using PipeT = decltype(Pipe::read());
+  static_assert(fpga_tools::has_subscript_v<PtrT>);
+  static_assert(std::is_same_v<PipeT, std::remove_const_t<std::remove_reference_t<decltype(std::declval<PtrT>()[0])>>>);
+
+  for (size_t i = 0; i < count_div_elements_per_cycle; i++) {
+    Pipe::write(in_ptr[i]);
+  }
+}
+
+template<typename Pipe, int elements_per_cycle, typename PtrT>
+void MemoryToPipe(PtrT in_ptr, size_t count_div_elements_per_cycle) {
+  using PipeT = decltype(Pipe::read());
+  static_assert(fpga_tools::has_subscript_v<PipeT>);
+  static_assert(fpga_tools::has_subscript_v<PtrT>);
+  static_assert(PipeT::size == elements_per_cycle);
+  static_assert(std::is_same_v<std::remove_reference_t<decltype(std::declval<PipeT>()[0])>, std::remove_const_t<std::remove_reference_t<decltype(std::declval<PtrT>()[0])>>>);
+
+  for (size_t i = 0; i < count_div_elements_per_cycle; i++) {
+    PipeT pipe_data;
+    #pragma unroll
+    for (int j = 0; j < elements_per_cycle; j++) {
+      pipe_data[j] = in_ptr[i * elements_per_cycle + j];
+    }
+    Pipe::write(pipe_data);
+  }
+}
+
+template<typename Pipe, typename PtrT>
+void StreamToPipe(PtrT out_ptr, size_t count_div_elements_per_cycle) {

```

Comment:
Shouldn't this be called PipeToMemory?

Reply:
Ya. I messed up. Not too worried about names right now.